{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "db57d478-68a1-449d-8369-4c80b63ef5e3",
   "metadata": {},
   "source": [
    "## **Usefull linux commands **\n",
    "\n",
    "---\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d903d1be-7af1-467d-9b59-24158577509f",
   "metadata": {},
   "source": [
    "• print name of current/working directory: <span style=\"color:purple\">pwd\n",
    "\n",
    "---\n",
    "\n",
    "• list directory contents: <span style=\"color:purple\">ls\n",
    "    \n",
    "---\n",
    "    \n",
    "• Make a  directory: <span style=\"color:purple\">mkdir mydirectory\n",
    "\n",
    "---\n",
    "\n",
    "• Go to a directory (file folder): <span style=\"color:purple\">cd mydirectory\n",
    "\n",
    "---\n",
    "\n",
    "•\tGo back to previous directory:<span style=\"color:purple\">cd ..\n",
    "\n",
    "---\n",
    "\n",
    "\n",
    "•\tLook into files: <span style=\"color:purple\">less file\n",
    "    \n",
    "---\n",
    "\n",
    "•\tLook into compressed (gz) files: <span style=\"color:purple\">zless file\n",
    "\n",
    "---\n",
    "\n",
    "•\tWrite files to screen: <span style=\"color:purple\">cat file\n",
    "\n",
    "---\n",
    "\n",
    "•\tWrite files to screen that are compressed (gz):  <span style=\"color:purple\">zcat\n",
    "\n",
    "---\n",
    "•\tTo look at the first 20 lines of a file:  <span style=\"color:purple\">head -n 20 file\n",
    "\n",
    "---\n",
    "•\tTo look at the end of a file:  <span style=\"color:purple\">tail\n",
    "\n",
    "---\n",
    "•\tCombine commands (pipe data): <span style=\"color:purple\"> |\n",
    "\n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e2dde6a-f734-4b74-bbf8-2dc6a3cbf593",
   "metadata": {},
   "source": [
    "### **Search for the information about the commands in the Mothur wiki page : https://www.mothur.org/wiki/  **"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36bae742-4371-483a-9a2b-812d9c4813f4",
   "metadata": {},
   "source": [
    "### • We will always work from the <span style=\"color:blue\">‘practical_course’ </span> folder in the  <span style=\"color:blue\">/home/ubuntu/  </span>home directory.\n",
    "Let’s see what are all the files we have there:\n",
    "\n",
    "---\n",
    "### 1. Let’s go through the different files we have.First, take a look at the output files from the Miseq machine:\n",
    "\n",
    "Undetermined_S0_L001_I1_001.fastq.gz\n",
    "\n",
    "Undetermined_S0_L001_R1_001.fastq.gz\n",
    "\n",
    "Undetermined_S0_L001_R2_001.fastq.gz\n",
    "\n",
    " <span style=\"color:red\">Q: Why do we have 3 files? What does R1, I1 and R2 in the files names mean? </span>.\n",
    "\n",
    "----\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "672269e7-9321-4035-bc86-3b6ecae6a2fa",
   "metadata": {},
   "source": [
    "Take a look at the first lines of the file Undetermined_S0_L001_I1_001.fastq.gz :\n",
    "\n",
    "<span style=\"color:purple\"> <br /> cd practical_course\n",
    "\n",
    "<span style=\"color:purple\"> cat Undetermined_S0_L001_I1_001.fastq.gz | head<br />\n",
    "\n",
    "\n",
    "@M03497:262:000000000-C5YFF:1:1101:11270:1550 1:N:0:0<br />\n",
    "CTTTTTCGCTTT<br />\n",
    "+<br />\n",
    "--,86;,+6,,9<br />\n",
    "\n",
    "@M03497:262:000000000-C5YFF:1:1101:10674:1560 1:N:0:0<br />\n",
    "CTCTTTTTTACT<br />\n",
    "+<br />\n",
    "6-8,,,,,+,,6<br />\n",
    "@M03497:262:000000000-C5YFF:1:1101:12013:1569 1:N:0:0<br />\n",
    "TTCTTTTCCTCT<br />\n",
    "\n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dcbc391-fb04-45f6-b1f1-6a8aa91a58ce",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "<span style=\"color:red\">Q: What is the first line?  <br />\n",
    "<span style=\"color:red\">What is the second line? <br />\n",
    "<span style=\"color:red\">What is the 4th line? <br />\n",
    "<span style=\"color:red\">What is the quality score of the first base called? Help:  <br/>\n",
    "    https://support.illumina.com/help/BaseSpace_OLH_009008/Content/Source/Informatics/BS/QualityScoreEncoding_swBS.htm <br/>\n",
    "<span style=\"color:red\">To what base-call accuracy does this correspond to? Help: <br/> https://en.wikipedia.org/wiki/Phred_quality_score\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c8224f8-5b6b-45c9-9078-a973adaaf800",
   "metadata": {},
   "source": [
    "2. Check the barcodes file where each barcode index is assigned to a sample and a locus (BV5=16S or PV9=18S).\n",
    "\n",
    "\n",
    "\n",
    "<span style=\"color:blue\"> practical_course_SamplesBarcodes_2019.xlsx\n",
    "    \n",
    " ![](pictures/1.png)\n",
    "\n",
    "\n",
    "This file was used to create the barcode file (<span style=\"color:blue\">barcodes_practical.txt </span>) that was used to search for the reads belonging to each sample/locus. This barcode file or “oligo file” is used to provide barcodes and primers to mothur.\n",
    "The format should be: barcode forwardBarcodeString reverseBarcodeString sampleName\n",
    "\n",
    "![](pictures/2.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a33f5f39-8b28-4c2f-afd2-07a4c07f218c",
   "metadata": {},
   "source": [
    "<span style=\"color:red\">Q: when we compare the first line the Index sequence is different in both files, why? Hint: how is the I read produced?<br/>\n",
    "<span style=\"color:red\">Q: Why is the second column “NONE”?\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ca440e7-9d93-4191-b2f6-2aa926d17127",
   "metadata": {
    "tags": []
   },
   "source": [
    "3. Now we can start processing the sequences. \n",
    "\n",
    "Start mothur from the  <span style=\"color:blue\">practical_course </span> (directory) in the terminal\n",
    "\n",
    "\n",
    "<span style=\"color:purple\"> <br />cd practical_course\n",
    "\n",
    "<span style=\"color:purple\"> <br />./mothur\n",
    "\n",
    "\n",
    "\n",
    "<span style=\"color:red\">Q: what version of mothur are we using?\n",
    "\n",
    "Normally you get the sequences from all samples together and you have to divide them this is called de-multiplexing. The first step is then to divide all the data into the different samples using the ‘fastq.info’ command.\n",
    "\n",
    "<span style=\"color:red\">Q: For this command we have to provide the oligos file why?\n",
    "    \n",
    "<span style=\"color:red\">Q: What does bdiffs=2 mean?\n",
    "    \n",
    "<span style=\"color:red\">Q: What is in the ‘File_FastqInfo.txt’ file? What information is it giving mothur?\n",
    "\n",
    "<span style=\"color:blue\"> <br />fastq.info(file=File_FastqInfo.txt, oligos=barcodes_practical.txt, bdiffs=2, fasta=F, qfile=F)\n",
    "\n",
    "#### <span style=\"color:darkblue\">Do not run this command! We already did it for you and the results are already in the folder. It’s the ‘File_FastqInfo...’ files.\n",
    "\n",
    "<span style=\"color:red\">Q: What were the different files produced? Why is there no Index file?\n",
    "\n",
    "This is an example of a sequence that was discarded in this step:\n",
    "\n",
    ">M03497_262_000000000-C5YFF_1_1101_11270_1550 | b\tfbdiffs=0(noMatch), rbdiffs=6(noMatch) fpdiffs=0(match), rpdiffs=0(match) \n",
    "AGAAGAAAGAAAAGAGGGAAGAAGAGAGTGGGGGGAGGGGTAGGGGGTGAGGGAGAGGGGGGGTGGTGAGGGGGAGGGGGAGGGGGGGGGGGGAGAAGGGGAGAGGAGATGGTTGGTGATGTAGTGGTGCGGGGGCGGGCGGGATCTGGGCCGGGCGGGGGTTGATAAGAGGTGGGGGATGCCCTTAGATGTTCTGGGCCGCACGCGCGCTNCNCTGNTGTGTTCAACGANTTNACACCNTGNCCGACAGGCCCGGGTAATCTTTGNANTTNNATCNTGNTGGGGATGGATCATTGNAATTTTTTGTTTTCCCGCAGAACTTCCTCCTCTTCCGCATCATCATCTCGCGTTTTCTCCCATCCCTCCCTTTCTTTCACACCCCCCTTCCCTCCTCCCACTTTCCTTCCTCTTTTCTTTTTTTTCCTCTCTTCTCCTTTTCTTCTCCTCCTCCCTCTCCATCTCTCCCCTTCCCCTCCCCCCTCCCC\n",
    "\n",
    "<span style=\"color:red\">Q: what does “rbdiffs” means? Why was this sequence trashed?\n",
    "\n",
    "\n",
    "This next step will pair single reads into paired-end reads and creates contigs.\n",
    "Let’s do this with the “make.contigs” command.\n",
    "\n",
    "<span style=\"color:blue\">make.contigs(file=Undetermined_S0_L001_R1_001.files, processors=8)\n",
    "\n",
    "<span style=\"color:red\">Q: What is in the ‘Undetermined_S0_L001_R1_001.files’ file? <br />\n",
    "<span style=\"color:red\">What information is this file giving to mothur?<br />\n",
    "<span style=\"color:red\">Q: How does mothur know which reads to pair together?<br />\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91bf794c-19b3-445d-9525-0c83867c74b5",
   "metadata": {},
   "source": [
    "4. Look at all the output files from the make.contigs command. \n",
    "\n",
    "•\tLook at the contigs report in the Undetermined_S0_L001_R1_001.contigs.report file\n",
    "\n",
    "![](pictures/4.png)\n",
    "\n",
    "<span style=\"color:red\">Q: what do the different columns mean? What is the main problem with the first contig? What do you think happened there?\n",
    "\n",
    "•\tThere are trim.contigs “fasta” and “qual” files produced. <br />\n",
    "<span style=\"color:red\">Q: What information is in these files? Why are there two types of files produced?\n",
    "\n",
    "•\tThere is a contigs.group file <br />\n",
    "<span style=\"color:red\">Q: what information is here?\n",
    "\n",
    "Let’s get a summary of the quality of our contigs. How much of the data we have generated might be useful data to analyse further.\n",
    "\n",
    "Here we use the “summary.seqs” command.\n",
    "\n",
    "<span style=\"color:blue\">summary.seqs(fasta= Undetermined_S0_L001_R1_001.trim.contigs.fasta)\n",
    "\n",
    "<span style=\"color:red\">Q: How many reads will we have to discard for further downstream analyses?\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f288b52-c107-43d1-99d0-7b4ec42191cc",
   "metadata": {},
   "source": [
    "5. Now that we have paired the reads in contigs \n",
    "\n",
    "(Undetermined_S0_L001_R1_001.trim.contigs.fasta)\n",
    "\n",
    "let’s filter those contigs to keep only those with really good quality using the “screen.seqs” command and the contigs report file.\n",
    "\n",
    "<span style=\"color:blue\">screen.seqs(fasta=Undetermined_S0_L001_R1_001.trim.contigs.fasta, contigsreport=Undetermined_S0_L001_R1_001.contigs.report, group=Undetermined_S0_L001_R1_001.contigs.groups, minoverlap=5, maxambig=0, maxhomop=10, minlength=100, maxlength=600, processors=8)\n",
    "\n",
    "<span style=\"color:red\">Q: How does this command work?\n",
    "    \n",
    "<span style=\"color:red\">Q: What do the different parameters mean? What is “maxambig”? \n",
    "\n",
    "<span style=\"color:red\">According to these factors will the first contig be trashed or kept? Why?\n",
    "\n",
    "![](pictures/5.png)\n",
    "\n",
    "<span style=\"color:red\">Q: How many sequences have been removed? Is this what we had expected?   \n",
    "<span style=\"color:red\">Q: what are the files produced? What are they? \n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3650098-a47f-4bb0-8ca2-8807a03a9f2d",
   "metadata": {},
   "source": [
    "6. We are now done with assembling the reads into good looking contigs. \n",
    "\n",
    "Now we have to separate the data because we have 18S and 16S PCR samples all mixed up. Because we know which samples are 18S and which are 16S (thanks to the barcodes/index) let’s rename the sequences accordingly.\n",
    "\n",
    "<span style=\"color:blue\">rename.seqs(fasta=Undetermined_S0_L001_R1_001.trim.contigs.good.fasta, group=Undetermined_S0_L001_R1_001.contigs.good.groups)\n",
    "\n",
    "<span style=\"color:red\">Q: How does this new group file looks like (Undetermined_S0_L001_R1_001.contigs.good.groups)? What information is there?\n",
    "\n",
    "Look at the headers of the new sequences in the renamed fasta file (Undetermined_S0_L001_R1_001.trim.contigs.good.renamed.fasta):\n",
    "\n",
    "     >1_B5.145\t\n",
    "TGCTCAAACCCCAAACATAGTGCGTACGCATACGCAGGTGTTACTGTGCAAATATTAGTAAATAGTCAATTATTGTTTGCGTTGTTACGTTGAAAGCGGTAATTGAGAGTGTTTGATATTCAACGGTACGCCTAACATCG\n",
    "    GGATCTACACCGAACGCAGGTTACTCACCGGTCTGAGATGCCCATTAGATGTTCTGGGCCGCACGCGCGCTACACTGATGTATTCAACGAGTTCACACCTTGGCCGACAGGCCCGGGTAATCTTTGAAATTTCATCGTGA\n",
    "    TGGGGATAGATCATTGCAATTGTTGGTCTACAACCAGGAAGTCCTAGTATTCGCCAGTCTTGCGCACGCGTTGACTACATCCCTTCCCTTTCTACACACCGACCGTCGCTCCTACCGATTGAATGATCAGGTGAAACGTT\n",
    "    CGGATCTCGGCAGCTTGTGTTGTTCCCCGCCCGGTACCTCGCACAAAGTCCACTAACACTTATGC\n",
    "\n",
    "<span style=\"color:red\">Q: what has changed? What info is given in the header?\n",
    "    \n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "225e6826-60e0-4888-8eae-a77167d4c29d",
   "metadata": {},
   "source": [
    "7. Now that we have renamed the sequences we can use a basic linux command (grep) to grab the sequences labeled “B5.” And those labeled “P9” to create separate files for the 16S and 18S samples. We do that with the fasta and the group file because we will need both of them later on. The output files are named BV5… (16S) or PV9… (18S)\n",
    "\n",
    "<span style=\"color:purple\">mothur>quit\n",
    "\n",
    "<span style=\"color:purple\">grep -A 1 \"B5.\" Undetermined_S0_L001_R1_001.trim.contigs.good.renamed.fasta>BV5.trim.contigs.good.renamed.fasta<br />\n",
    "    \n",
    "<span style=\"color:purple\">grep \"B5.\" Undetermined_S0_L001_R1_001.contigs.good.renamed.groups>BV5.contigs.good.renamed.groups<br />\n",
    "    \n",
    "<span style=\"color:purple\">grep -A 1 \"P9.\" Undetermined_S0_L001_R1_001.trim.contigs.good.renamed.fasta>PV9.trim.contigs.good.renamed.fasta<br />\n",
    "    \n",
    "<span style=\"color:purple\">grep \"P9.\" Undetermined_S0_L001_R1_001.contigs.good.renamed.groups>PV9.contigs.good.renamed.groups<br />\n",
    "\n",
    "\n",
    "\n",
    "### These next steps concern the analysis of the 16S samples that we will analyze together. The 18S samples you will analyze on your own :D\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a2ea200-4169-47aa-81bf-e202f1f24862",
   "metadata": {},
   "source": [
    "---\n",
    "8. Before moving on let’s reduce the size of the data to reduce the computing time. For this let’s “unique” our sequences. In this process, if two identical sequences are found one will be removed from the fasta file therefore listing only unique sequences in the fasta file.\n",
    "\n",
    "Back to mothur:\n",
    "\n",
    "<span style=\"color:purple\">./mothur\n",
    "\n",
    "<span style=\"color:blue\">unique.seqs(fasta=BV5.trim.contigs.good.renamed.fasta)\n",
    "\n",
    "<span style=\"color:red\">Q: what are the files produced? \n",
    "\n",
    "Look into the BV5.trim.contigs.good.renamed.names\n",
    "\n",
    "One line looks like this:\n",
    "\n",
    "![](pictures/8.png)\n",
    "\n",
    "<span style=\"color:red\">Q: This is a list of sequence names. What do you think this can mean? How is this linked to the fasta file containing the unique sequences\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a58b7e1d-12a6-4a1d-a611-56aba8252264",
   "metadata": {},
   "source": [
    "9. As you just saw this “names” files is quite complicated so let’s transform it into a “count” file.\n",
    "\n",
    "<span style=\"color:blue\">count.seqs(name=BV5.trim.contigs.good.renamed.names, group=BV5.contigs.good.renamed.groups)\n",
    "\n",
    "check the BV5.trim.contigs.good.renamed.count_table file produced. If pasted into excel the first lines and columns would look like this:\n",
    "\n",
    "![](pictures/9.png)\n",
    "    \n",
    "<span style=\"color:red\">Q: What do you think this can mean? How is this linked to the fasta file containing the unique sequences?\n",
    "\n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6178b33-abae-4267-8cfe-1fb63cd10c04",
   "metadata": {},
   "source": [
    "10. OK let’s keep the count file and move-on to chimera detection. \n",
    "\n",
    "<span style=\"color:blue\">chimera.uchime(fasta=BV5.trim.contigs.good.renamed.unique.fasta, count=BV5.trim.contigs.good.renamed.count_table, processors=8)\n",
    "\n",
    "<span style=\"color:red\">Q: What are chimeras? How/when are they produced?\n",
    "\n",
    "Check the Mothur wiki and see what are the parameters for chimera detection.\n",
    "https://www.mothur.org/wiki/Chimera.uchime\n",
    "\n",
    "Check the Uchime page to understand how it works https://drive5.com/usearch/manual/chimeras.html\n",
    "https://drive5.com/usearch/manual/otu_qc_chimeras.html   \n",
    "#### <span style=\"color:green\">(end of day 1 ~ 5 h)\n",
    "\n",
    "\n",
    "<span style=\"color:red\">Q: What are the output files?\n",
    "\n",
    "Let’s check the chimera file:\n",
    "\n",
    "![](pictures/10.png)\n",
    "\n",
    "For info check here https://www.drive5.com/usearch/manual8.1/uchimeout.html and https://www.drive5.com/usearch/manual8.1/UCHIME_score.html\n",
    "\n",
    "<span style=\"color:red\">Q: Is the 8th sequence a chimera?\n",
    "    \n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3be220c7-a124-468a-8bbe-f310120bd4b7",
   "metadata": {},
   "source": [
    "11. Let’s now remove all the identified chimeric sequences from our data.\n",
    "\n",
    "<span style=\"color:blue\">remove.seqs(accnos=BV5.trim.contigs.good.renamed.unique.denovo.uchime.accnos, fasta=BV5.trim.contigs.good.renamed.unique.fasta, count=BV5.trim.contigs.good.renamed.count_table)\n",
    "\n",
    "<span style=\"color:red\">Q: How many reads did we loose?\n",
    "\n",
    "    \n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31b71bc8-93d3-49b4-b8a0-a722c44aa050",
   "metadata": {},
   "source": [
    "![](pictures/f1.jpg)\n",
    "    \n",
    "---    \n",
    "    \n",
    "12. Now let’s identify the reads that we have by doing a taxonomical classification of each read against a data base. But first let’s look into the database so we’ll understand how it is constructed.\n",
    "\n",
    "For the 16S the data base is the 16S green genes (gg) data base http://greengenes.secondgenome.com\n",
    "\n",
    "For the 18S is the PR2 data base https://github.com/pr2database/pr2database/releases\n",
    "\n",
    "Let’s look into the gg data base:\n",
    "\n",
    "The fasta file (gg_13_8_99_phiX.fasta) looks like a normal fasta file, the first sequence is this:\n",
    "\n",
    "    >1111886\n",
    "\n",
    "AACGAACGCTGGCGGCATGCCTAACACATGCAAGTCGAACGAGACCTTCGGGTCTAGTGGCGCACGGGTGCGTAACGCGTGGGAATCTGCCCTTGGGTACGGAATAACAGTTAGAAATGACTGCTAATACCGTATAATGA<br />\n",
    "CTTCGGTCCAAAGATTTATCGCCCAGGGATGAGCCCGCGTAGGATTAGCTTGTTGGTGAGGTAAAGGCTCACCAAGGCGACGATCCTTAGCTGGTCTGAGAGGATGATCAGCCACACTGGGACTGAGACATGGCCCAGAC<br />\n",
    "TCCTACGGGAGGCAGCAGTGGGGAATATTGGACAATGGGCGAAAGCCTGATCCAGCAATGCCGCGTGAGTGATGAAGGCCTTAGGGTTGTAAAGCTCTTTTACCCGGGATGATAATGACAGTACCGGGAGAATAAGCCCC<br />GGCTAACTCCGTGCCAGCAGCCGCGGTAATACGGAGGGGGCTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGCACGTAGGCGGCTTTGTAAGTTAGAGGTGAAAGCCCGGGGCTCAACTCCGGAATTGCCTTTAAGA<br />CTGCATCGCTAGAATTGTGGAGAGGTGAGTGGAATTCCGAGTGTAGAGGTGAAATTCGTAGATATTCGGAAGAACACCAGTGGCGAAGGCGACTCACTGGACACATATTGACGCTGAGGTGCGAAAGCGTGGGGAGCA<br />AACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGATGACTAGCTGTCGGGGCGCTTAGCGTTTCGGTGGCGCAGCTAACGCGTTAAGTCATCCGCCTGGGGAGTACGGCCGCAAGGTTAAAACTCAAAGAA<br />ATTGACGGGGGCCTGCACAAGCGGTGGAGCATGTGGTTTAATTCGAAGCAACGCGCAGAACCTTACCAGCGTTTGACATGGTAGGACGGTTTCCAGAGATGGATTCCTACCCTTACGGGACCTACACACAGGTGCTGC<br />ATGGCTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTCGTCTTTGGTTGCTACCATTTAGTTGAGCACTCTAAAAAAACTGCCGGTGATAAGCCGGAGGAAGGTGGGGATGACG<br />TCAAGTCCTCATAGCCCTTACGCGCTGGGCTACACACGTGCTACAATGGCGGTGACAGAGGGCAGCAAACCCGCGAGGGTGAGCTAATCTCCAAAAGCCGTCTCAGTTCGGATTGTTCTCTGCAACTCGAGAGCATGA<br />AGGCGGAATCGCTAGTAATCGCGGATCAGCACGCCGCGGTGAATACGTTCCCAGGCCTTGTACACACCGCCCGTCACATCACGAAAGTCGGTTGCACTAGAAGTCGGTGGGCTAACCCGCAAGGGAGGCAGCCGCCTA<br />AAGTGTGATCGGTAATTGGGGTG\n",
    "\n",
    "Now let’s look the tax file (gg_13_8_99_phiX.gg.tax). The first entry is this:\n",
    "\n",
    "4\tk__Archaea;p__[Parvarchaeota];c__[Parvarchaea];o__YLA114;\n",
    "\n",
    "This entry gives the taxonomical classification of sequence “4” in the fasta file. k,p,c,o indicate the classification at the kingdom, phylum, class and order level.\n",
    "\n",
    "Lets’ try to find in the tax file, the taxonomic classification of the first sequence in the fasta (sequence “1111886”).\n",
    "\n",
    "<span style=\"color:purple\">grep \"1111886\" gg_13_8_99_phiX.gg.tax\n",
    "\n",
    "\n",
    "The output is:\n",
    "\n",
    "     >1111886\t\n",
    "   \n",
    "     k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Kaistobacter;\n",
    "\n",
    "So the first sequence in the fasta file comes from a strain of the Kaistobacter genus.\n",
    "\n",
    "<span style=\"color:red\">Q: The second sequence in the fasta i.e. sequence “1111883”, from which organism is it? What is the taxonomical classification associated?\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fd4f41f-dbf1-4cdb-b860-9f48eb78ff95",
   "metadata": {},
   "source": [
    "13. Now let’s use the 16S data base to classify our sequences using the “classify.seqs” command.\n",
    "\n",
    "<span style=\"color:blue\">classify.seqs(fasta=BV5.trim.contigs.good.renamed.unique.pick.fasta, template=gg_13_8_99_phiX.fasta, taxonomy=gg_13_8_99_phiX.gg.tax, processors=2)\n",
    "\n",
    "Let’s look how the classify.seqs command works  https://www.mothur.org/wiki/Classify.seqs\n",
    "\n",
    "<span style=\"color:red\">Q: What are the default parameters?\n",
    "    \n",
    "<span style=\"color:red\">Q: What are “Kmers”? What “Kmer” size are we using? How does the RDP classifier works? What is “bootstrapping” and how does it work?\n",
    "    \n",
    "<span style=\"color:red\">Q: What are the other possibilities for classification? How could they be better?\n",
    "    \n",
    "\n",
    "Let’s look the taxonomy file which is produced: BV5.trim.contigs.good.renamed.unique.pick.gg.wang.taxonomy\n",
    "\n",
    "1_B5.145\tunknown;unknown_unclassified(0);unknown_unclassified(0);unknown_unclassified(0);unknown_unclassified(0);unknown_unclassified(0);unknown_unclassified(0);\n",
    "2_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rickettsiales(100);f__mitochondria(100);f__mitochondria_unclassified(100);f__mitochondria_unclassified(100);\n",
    "3_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rickettsiales(100);f__mitochondria(100);g__Arabidopsis(100);s__thaliana(100);\n",
    "4_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rickettsiales(100);f__mitochondria(100);f__mitochondria_unclassified(100);f__mitochondria_unclassified(100);\n",
    "5_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Pseudomonadales(100);f__Pseudomonadaceae(100);g__Pseudomonas(100);g__Pseudomonas_unclassified(100);\n",
    "6_B5.145\tk__Bacteria(100);p__Cyanobacteria(100);c__Chloroplast(100);o__Streptophyta(100);o__Streptophyta_unclassified(100);o__Streptophyta_unclassified(100);o__Streptophyta_unclassified(100);\n",
    "7_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Xanthomonadales(100);f__Xanthomonadaceae(100);g__Stenotrophomonas(99);g__Stenotrophomonas_unclassified(99);\n",
    "8_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Betaproteobacteria(100);o__Burkholderiales(100);f__Comamonadaceae(100);f__Comamonadaceae_unclassified(100);f__Comamonadaceae_unclassified(100);\n",
    "9_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rickettsiales(100);f__mitochondria(100);g__Arabidopsis(88);s__thaliana(88);\n",
    "10_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Pseudomonadales(100);f__Moraxellaceae(100);g__Acinetobacter(100);s__rhizosphaerae(87);\n",
    "\n",
    "<span style=\"color:red\">Q: What are the values in parenthesis? How were they calculated?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "872a6219-5bfb-4895-9447-1e43af8295b3",
   "metadata": {},
   "source": [
    "13. Now let’s use the 16S data base to classify our sequences using the “classify.seqs” command.\n",
    "\n",
    "<span style=\"color:blue\">classify.seqs(fasta=BV5.trim.contigs.good.renamed.unique.pick.fasta, template=gg_13_8_99_phiX.fasta, taxonomy=gg_13_8_99_phiX.gg.tax, processors=2)\n",
    "\n",
    "Let’s look how the classify.seqs command works  https://www.mothur.org/wiki/Classify.seqs\n",
    "\n",
    "<span style=\"color:red\">Q: What are the default parameters?\n",
    "    \n",
    "<span style=\"color:red\">Q: What are “Kmers”? What “Kmer” size are we using?How does the RDP classifier works? What is “bootstrapping” and how does it work?\n",
    "    \n",
    "<span style=\"color:red\">Q: What are the other possibilities for classification? How could they be better?\n",
    "\n",
    "Let’s look the taxonomy file which is produced: BV5.trim.contigs.good.renamed.unique.pick.gg.wang.taxonomy\n",
    "\n",
    "1_B5.145\tunknown;unknown_unclassified(0);unknown_unclassified(0);unknown_unclassified(0);unknown_unclassified(0);unknown_unclassified(0);unknown_unclassified(0);\n",
    "2_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rickettsiales(100);f__mitochondria(100);f__mitochondria_unclassified(100);f__mitochondria_unclassified(100);\n",
    "3_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rickettsiales(100);f__mitochondria(100);g__Arabidopsis(100);s__thaliana(100);\n",
    "4_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rickettsiales(100);f__mitochondria(100);f__mitochondria_unclassified(100);f__mitochondria_unclassified(100);\n",
    "5_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Pseudomonadales(100);f__Pseudomonadaceae(100);g__Pseudomonas(100);g__Pseudomonas_unclassified(100);\n",
    "6_B5.145\tk__Bacteria(100);p__Cyanobacteria(100);c__Chloroplast(100);o__Streptophyta(100);o__Streptophyta_unclassified(100);o__Streptophyta_unclassified(100);o__Streptophyta_unclassified(100);\n",
    "7_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Xanthomonadales(100);f__Xanthomonadaceae(100);g__Stenotrophomonas(99);g__Stenotrophomonas_unclassified(99);\n",
    "8_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Betaproteobacteria(100);o__Burkholderiales(100);f__Comamonadaceae(100);f__Comamonadaceae_unclassified(100);f__Comamonadaceae_unclassified(100);\n",
    "9_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rickettsiales(100);f__mitochondria(100);g__Arabidopsis(88);s__thaliana(88);\n",
    "10_B5.145\tk__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Pseudomonadales(100);f__Moraxellaceae(100);g__Acinetobacter(100);s__rhizosphaerae(87);\n",
    "\n",
    "<span style=\"color:red\">Q: What are the values in parenthesis? How were they calculated?\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca45f031-ae9b-4c19-a497-aacbfefdb895",
   "metadata": {},
   "source": [
    "14. Now that we have (individually) classified all the 16S reads we had let’s cluster the reads into groups (clusters) based on sequence similarity. These groups are called OTUs for Operational Taxonomical Units, a term used to describe technically defined microbial taxonomical groups.\n",
    "\n",
    "<span style=\"color:blue\">cluster(fasta=BV5.trim.contigs.good.renamed.unique.pick.fasta, count=BV5.trim.contigs.good.renamed.pick.count_table, method=dgc, cutoff=0.03)\n",
    "\n",
    "Let’s check the “cluster” command https://www.mothur.org/wiki/Cluster\n",
    "\n",
    "<span style=\"color:red\">Q: What are the default parameters? Which method are we using? What distance are we using? What does it mean in terms of sequence similarity?\n",
    "    \n",
    "<span style=\"color:red\">Q: What are the other methods for clustering? How are they different?\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94b809f8-0c5a-4658-850c-9491e8e30077",
   "metadata": {},
   "source": [
    "15. The last command grouped the sequences into different OTUs creating a “.list” file. But what do we do with OTUs which include only 1 or 10 reads? Is this meaningful or are these artefacts? We propose to keep only the OTUs with more than 50 reads. For this we do abundance filtering.\n",
    "\n",
    "<span style=\"color:blue\">split.abund(fasta=BV5.trim.contigs.good.renamed.unique.pick.fasta, count=BV5.trim.contigs.good.renamed.pick.count_table, list=BV5.trim.contigs.good.renamed.unique.pick.dgc.list, cutoff=50)\n",
    "\n",
    "<span style=\"color:red\">Q: What files are created? What do they contain?\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d821eb67-850c-4087-a04d-315f6130d32e",
   "metadata": {},
   "source": [
    "15. The last command grouped the sequences into different OTUs creating a “.list” file. But what do we do with OTUs which include only 1 or 10 reads? Is this meaningful or are these artefacts? We propose to keep only the OTUs with more than 50 reads. For this we do abundance filtering.\n",
    "\n",
    "split.abund(fasta=BV5.trim.contigs.good.renamed.unique.pick.fasta, count=BV5.trim.contigs.good.renamed.pick.count_table, list=BV5.trim.contigs.good.renamed.unique.pick.dgc.list, cutoff=50)\n",
    "\n",
    "<span style=\"color:red\">Q: What files are created? What do they contain\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ee27ccc-6334-47f8-8308-d3ce1ed89d58",
   "metadata": {},
   "source": [
    "16. On step 13 we classified all of the sequences individually which created a taxonomy file giving the classification of each sequence (BV5.trim.contigs.good.renamed.unique.pick.gg.wang.taxonomy). \n",
    "Then, on step 15 we grouped sequences into OTUs. In this step we will combine both outputs to give a taxonomical classification to the OTUs using the taxonomical classification of the individual sequences.\n",
    "\n",
    "<span style=\"color:blue\">classify.otu(taxonomy=BV5.trim.contigs.good.renamed.unique.pick.gg.wang.taxonomy, list=BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.list, count=BV5.trim.contigs.good.renamed.pick.0.03.abund.count_table)\n",
    "\n",
    "Check the command info https://www.mothur.org/wiki/Classify.otu\n",
    "\n",
    "<span style=\"color:red\">Q: What are the default parameters? Which method are we using?\n",
    "<span style=\"color:red\">Q: What files are created? What do they contain?\n",
    "\n",
    "The consensus taxonomy file (cons.taxonomy) gives the consensus taxonomy of each OTU. It looks like this:\n",
    "\n",
    "OTU\tSize\tTaxonomy\n",
    "Otu00001\t60456\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rhodospirillales(100);f__Acetobacteraceae(100);g__Acetobacter(100);g__Acetobacter_unclassified(100);\n",
    "Otu00002\t46355\tk__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Lactobacillales(100);f__Lactobacillaceae(100);g__Lactobacillus(100);s__helveticus(76);\n",
    "Otu00003\t13332\tk__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Enterobacteriales(100);f__Enterobacteriaceae(100);f__Enterobacteriaceae_unclassified(61);f__Enterobacteriaceae_unclassified(61);\n",
    "Otu00004\t7155\tk__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Pseudomonadales(100);f__Pseudomonadaceae(100);g__Pseudomonas(100);g__Pseudomonas_unclassified(64);\n",
    "Otu00005\t7152\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rhodospirillales(100);f__Acetobacteraceae(100);g__Acetobacter(100);g__Acetobacter_unclassified(100);\n",
    "Otu00006\t6700\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rickettsiales(100);f__mitochondria(100);f__mitochondria_unclassified(100);f__mitochondria_unclassified(100);\n",
    "Otu00007\t5399\tk__Bacteria(100);p__Fusobacteria(100);c__Fusobacteriia(100);o__Fusobacteriales(100);f__Fusobacteriaceae(100);g__Fusobacterium(100);g__Fusobacterium_unclassified(100);\n",
    "Otu00008\t5022\tk__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Lactobacillales(100);f__Lactobacillaceae(100);g__Lactobacillus(97);s__vaginalis(53);\n",
    "Otu00009\t4943\tk__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Lactobacillales(100);f__Enterococcaceae(100);g__Enterococcus(100);s__cecorum(100);\n",
    "Otu00010\t4308\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rickettsiales(100);f__Rickettsiaceae(100);g__Wolbachia(100);g__Wolbachia_unclassified(100);\n",
    "Otu00011\t3818\tk__Bacteria(100);p__Cyanobacteria(100);c__Chloroplast(100);o__Streptophyta(100);o__Streptophyta_unclassified(100);o__Streptophyta_unclassified(100);o__Streptophyta_unclassified(100);\n",
    "Otu00012\t2725\tk__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Enterobacteriales(100);f__Enterobacteriaceae(100);g__Pantoea(81);g__Pantoea_unclassified(81);\n",
    "Otu00013\t2722\tk__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Lactobacillales(100);f__Lactobacillaceae(100);g__Lactobacillus(100);s__plantarum(99);\n",
    "Otu00014\t2583\tk__Bacteria(100);p__Proteobacteria(100);c__Betaproteobacteria(100);o__Burkholderiales(100);f__Comamonadaceae(100);f__Comamonadaceae_unclassified(96);f__Comamonadaceae_unclassified(96);\n",
    "Otu00015\t2150\tk__Bacteria(100);p__Proteobacteria(100);c__Betaproteobacteria(100);o__Burkholderiales(100);f__Comamonadaceae(100);g__Delftia(100);g__Delftia_unclassified(100);\n",
    "Otu00016\t2003\tk__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Lactobacillales(100);f__Lactobacillaceae(100);g__Lactobacillus(100);g__Lactobacillus_unclassified(65);\n",
    "Otu00017\t1725\tk__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Pseudomonadales(100);f__Pseudomonadaceae(100);g__Pseudomonas(100);g__Pseudomonas_unclassified(51);\n",
    "Otu00018\t1589\tk__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Lactobacillales(100);f__Enterococcaceae(100);g__Enterococcus(99);g__Enterococcus_unclassified(95);\n",
    "Otu00019\t1519\tk__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Bacillales(100);f__Bacillaceae(100);g__Bacillus(90);g__Bacillus_unclassified(90);\n",
    "\n",
    "<span style=\"color:red\">Q: What do the numbers in parenthesis mean? How were they calculated?\n",
    "    \n",
    "<span style=\"color:red\">Q: What cutoff did we use? Do you think it is rather high and conservative or low and not so conservative?\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5adc3572-3d78-46e3-ba06-dd5e8be080c7",
   "metadata": {},
   "source": [
    "17. All these list files and count files are a bit too complicated… So now we want to create a simple file indicating the number of sequences per OTU and per sample (OTU table). For this we use the share command to create a share file using the abundance filtered “list” and “count” files from step 15. \n",
    "\n",
    "<span style=\"color:blue\">make.shared(list=BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.list, count=BV5.trim.contigs.good.renamed.pick.0.03.abund.count_table)\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f459bbd1-19c2-47eb-8200-0faf533ddce6",
   "metadata": {},
   "source": [
    "18. So now we have a nice OTU table we can work with! This table is associated to the cons.taxonomy file created in step 16 (BV5.trim.contigs.good.renamed.unique.pick.dgc.unique_list.0.03.abund.0.03.cons.taxonomy) which gives the OTU taxonomical classification. Now we want to retrieve the fasta sequences associated to each of these OTUs. Because each OTU is constituted of up to 13K sequences it makes no sense to retrieve all of them, so we’ll retrieve OTU representative sequences.\n",
    "\n",
    "<span style=\"color:blue\">get.oturep(fasta=BV5.trim.contigs.good.renamed.unique.pick.0.03.abund.fasta, list=BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.list, count=BV5.trim.contigs.good.renamed.pick.0.03.abund.count_table, method=abundance)\n",
    "\n",
    "Lets’ see how the command works https://www.mothur.org/wiki/Get.oturep\n",
    "\n",
    "<span style=\"color:red\">Q: How are these OTU representative sequences chosen?\n",
    "    \n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73624e68-b03c-4ad0-8309-1d971f613bb2",
   "metadata": {},
   "source": [
    "19. Let’s explore the OTU table\n",
    "\n",
    "Let’s open this shared file BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.shared in excel to see how it looks:\n",
    "\n",
    "![](pictures/19.png)\n",
    "\n",
    "The first column indicates the clustering distance.\n",
    "The second column indicates the name of the sample based on the barcode id. So the first three samples are Arabidopsis leaves, then the next three samples are soil, then the next three samples are chicken poo and the next three samples are fruit flies.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "980845b4-d646-4dd7-a16f-30f2092db7a5",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "19.1. One of the most abundant OTUs in Arabidopsis leaves is OTU 11.\n",
    "Let’s check the taxonomical classification of this OTU:\n",
    "\n",
    "<span style=\"color:purple\">head -n 20 BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.0.03.cons.taxonomy\n",
    "\n",
    "\n",
    "The output looks like:\n",
    " \n",
    "OTU\tSize\tTaxonomy\n",
    "    \n",
    "Otu00001\t60456\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rhodospirillales(100);f__Acetobacteraceae(100);g__Acetobacter(100);g__Acetobacter_unclassified(100);\n",
    "Otu00002\t46355\n",
    "    k__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Lactobacillales(100);f__Lactobacillaceae(100);g__Lactobacillus(100);s__helveticus(76);\n",
    "    \n",
    "Otu00003\t13332\t\n",
    "    k__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Enterobacteriales(100);f__Enterobacteriaceae(100);f__Enterobacteriaceae_unclassified(61);f__Enterobacteriaceae_unclassified(61);\n",
    "    \n",
    "Otu00004\t7155\n",
    "    k__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Pseudomonadales(100);f__Pseudomonadaceae(100);g__Pseudomonas(100);g__Pseudomonas_unclassified(64);\n",
    "Otu00005\t7152\t\n",
    "    k__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rhodospirillales(100);f__Acetobacteraceae(100);g__Acetobacter(100);g__Acetobacter_unclassified(100);\n",
    "Otu00006\t6700\t\n",
    "    k__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rickettsiales(100);f__mitochondria(100);f__mitochondria_unclassified(100);f__mitochondria_unclassified(100);\n",
    "Otu00007\t5399\t\n",
    "    k__Bacteria(100);p__Fusobacteria(100);c__Fusobacteriia(100);o__Fusobacteriales(100);f__Fusobacteriaceae(100);g__Fusobacterium(100);g__Fusobacterium_unclassified(100);\n",
    "Otu00008\t5022\n",
    "    k__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Lactobacillales(100);f__Lactobacillaceae(100);g__Lactobacillus(97);s__vaginalis(53);\n",
    "Otu00009\t4943\tk__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Lactobacillales(100);f__Enterococcaceae(100);g__Enterococcus(100);s__cecorum(100);\n",
    "Otu00010\t4308\tk__Bacteria(100);p__Proteobacteria(100);c__Alphaproteobacteria(100);o__Rickettsiales(100);f__Rickettsiaceae(100);g__Wolbachia(100);g__Wolbachia_unclassified(100);\n",
    "    \n",
    "<span style=\"color:green\">Otu00011\t3818\n",
    "    \n",
    "<span style=\"color:green\">k__Bacteria(100);p__Cyanobacteria(100);c__Chloroplast(100);o__Streptophyta(100);o__Streptophyta_unclassified(100);o__Streptophyta_unclassified(100);o__Streptophyta_unclassified(100);\n",
    "    \n",
    "Otu00012\t2725\tk__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Enterobacteriales(100);f__Enterobacteriaceae(100);g__Pantoea(81);g__Pantoea_unclassified(81);\n",
    "Otu00013\t2722\tk__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Lactobacillales(100);f__Lactobacillaceae(100);g__Lactobacillus(100);s__plantarum(99);\n",
    "Otu00014\t2583\tk__Bacteria(100);p__Proteobacteria(100);c__Betaproteobacteria(100);o__Burkholderiales(100);f__Comamonadaceae(100);f__Comamonadaceae_unclassified(96);f__Comamonadaceae_unclassified(96);\n",
    "Otu00015\t2150\tk__Bacteria(100);p__Proteobacteria(100);c__Betaproteobacteria(100);o__Burkholderiales(100);f__Comamonadaceae(100);g__Delftia(100);g__Delftia_unclassified(100);\n",
    "Otu00016\t2003\tk__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Lactobacillales(100);f__Lactobacillaceae(100);g__Lactobacillus(100);g__Lactobacillus_unclassified(65);\n",
    "Otu00017\t1725\tk__Bacteria(100);p__Proteobacteria(100);c__Gammaproteobacteria(100);o__Pseudomonadales(100);f__Pseudomonadaceae(100);g__Pseudomonas(100);g__Pseudomonas_unclassified(51);\n",
    "Otu00018\t1589\tk__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Lactobacillales(100);f__Enterococcaceae(100);g__Enterococcus(99);g__Enterococcus_unclassified(95);\n",
    "Otu00019\t1519\tk__Bacteria(100);p__Firmicutes(100);c__Bacilli(100);o__Bacillales(100);f__Bacillaceae(100);g__Bacillus(90);g__Bacillus_unclassified(90);\n",
    "\n",
    "Q: What is this OTU3? Is it common to find Cyanobacteria on plant leaves? Keep in mind that Arabidopsis is not an aquatic plant… \n",
    "\n",
    "Let’s get the OTU representative sequence and blast it to verify the identity of this OTU:\n",
    "\n",
    "<span style=\"color:purple\">head -n 25 BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.0.03.rep.fasta\n",
    "\n",
    "    >6_B5.145\tOtu00011|3818|B5.145-B5.146-B5.147-B5.148-B5.149-B5.151-B5.152-B5.153-B5.154-B5.155-B5.156\n",
    "    \n",
    "    \n",
    "CAGTAGTCCTAGCCGTAAACGATGGATACTAGGCGCTGTGCGTATCGACCCGTGCAGTGCTGTAGCTAACGCGTTAAGTATCCCGCCTGGGGAGTACGTTCGCAAGAATGAAACTCAAAGGAATTGACGGGGGCCCGCAC\n",
    "    AAGCGGTGGAGCATGTGGTTTAATTCGATGCAAAGCGAAGAACCTTACCAGGGCTTGACATGCCGCGAATCCTCTTGAAAGAGAGGGGTGCCTTCGGGAACGCGGACACAGGTGGTGCATGGCTGTCGTCAGCTCGTGCC\n",
    "    GTAAGGTGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTCGTGTTTAGTTGCCACCGTTGAGTTTGGAACCCTGAACAGACTGCCGGTGATAAGCCGGA\n",
    "\n",
    "<span style=\"color:red\">Q: What is blast?  Have you used it? What is the nr database?\n",
    "\n",
    "Let’s blast this sequence against the nr databse of the NCBI. https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome\n",
    "\n",
    "This is the output you should get. It looks like this sequence matches something else…\n",
    "\n",
    "\n",
    "![](pictures/19.1.png)\n",
    "\n",
    "<span style=\"color:red\">Q: What is the real identity of this OTU? Are these cyanobacteria? What did we do wrong? What should we do with this OTU?\n",
    "<span style=\"color:red\">Q: Why don’t we find this OTU in the other samples except chicken poo (in high abundance)?\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d00fa4f1-37c8-4bcf-9b35-f8c68b4eb627",
   "metadata": {},
   "source": [
    "19.2. The most abundant OTU in chicken poo is OTU 2.\n",
    "\n",
    "<span style=\"color:red\">Q: What is the identity of this OTU? Does it make sense to find this kind of bacteria in chicken poo? Can you confirm the identity of this OTU? What should we do with this OTU?\n",
    "\n",
    "---\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d4f3e32-7aa2-4196-b4e0-25b578bd481b",
   "metadata": {
    "tags": []
   },
   "source": [
    "19.3. What is this OTU6 in Arabidopsis leaves?\n",
    "\n",
    "<span style=\"color:red\">Q: What is the identity of this OTU? Does it make sense to find Rickettsiales in Arabidopsis leaves? Can you confirm the identity of this OTU? What should we do with this OTU?\n",
    "    \n",
    "<span style=\"color:red\">Q: Why don’t we find this OTU in the other samples?\n",
    "\n",
    "We seem to have some problems so let’s look more into detail in the taxonomy file to see what else looks weird.\n",
    "\n",
    "<span style=\"color:purple\">head -n 50 BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.0.03.cons.taxonomy\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5425f7e3-4dff-42a8-aad4-7290339239d6",
   "metadata": {},
   "source": [
    "19.4. One other OTU looks weird: OTU23\n",
    "\n",
    "Otu00023\t1225\tunknown(100);unknown_unclassified(100);unknown_unclassified(100);unknown_unclassified(100);unknown_unclassified(100);unknown_unclassified(100);unknown_unclassified(100);\n",
    "\n",
    "<span style=\"color:red\">Q: Did we just discovered a new microbial taxon?! Can you confirm that this OTU is really unknown? Why do we find this OTU in all the samples? What should we do with this OTU? \n",
    "    \n",
    "---\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7efa34e1-1718-4339-9ac0-3d3db92cea3d",
   "metadata": {},
   "source": [
    "19.5. One other OTU looks weird: OTU28\n",
    "\n",
    "Otu00034795\t\n",
    "PhiX(64);PhiX_unclassified(64);PhiX_unclassified(64);PhiX_unclassified(64);PhiX_unclassified(64);PhiX_unclassified(64);PhiX_unclassified(64);\n",
    "\n",
    "<span style=\"color:red\">Q: Phi X are a type of bacterio phages, mostly infecting E. coli.  Why are we amplifying bacteriphages DNA: do bacteriophages have 16S? \n",
    "<span style=\"color:red\">Q: Can you confirm that this OTU is really from this bacteriophage? Why do we find this OTU in all the samples? What should we do with this OTU?\n",
    "\n",
    "---    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62e69ebf-8a90-4751-bf25-3fbfab5ba250",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "20. This is the last step of the data processing (finaly!). \n",
    "\n",
    "![](pictures/f2.jpg)\n",
    "\n",
    "\n",
    "\n",
    "Here we will clean the data more thoroughly to get rid of the remaining artifacts. We use the remove lineage command to get rid of non-target ‘lineages’ and taxonomical groups.\n",
    "\n",
    "<span style=\"color:blue\">remove.lineage(constaxonomy=BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.0.03.cons.taxonomy, shared=BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.shared, taxon=unknown-mitochondria-Chloroplast-PhiX)\n",
    "\n",
    "<span style=\"color:blue\">remove.lineage(constaxonomy=BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.0.03.cons.pick.taxonomy, shared=BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.0.03.pick.shared, taxon='mitochondria-Chloroplast ')\n",
    "\n",
    "<span style=\"color:blue\">remove.lineage(constaxonomy=BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.0.03.cons.pick.pick.taxonomy, shared=BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.0.03.pick.0.03.pick.shared, taxon='Chloroplast')\n",
    "\n",
    "<span style=\"color:blue\">remove.lineage(constaxonomy=BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.0.03.cons.pick.pick.pick.taxonomy, shared=BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.0.03.pick.0.03.pick.0.03.pick.shared, taxon='PhiX')\n",
    "\n",
    "<span style=\"color:red\">Q: Can you explain what are the different “lineages” removed? And why they were removed?\n",
    "\n",
    "\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed0d1300-706c-4319-a062-7215f8cc7925",
   "metadata": {},
   "outputs": [],
   "source": [
    "17. All these list files and count files are a bit too complicated… So now we want to create a simple file indicating the number of sequences per OTU and per sample (OTU table). For this we use the share command to create a share file using the abundance filtered “list” and “count” files from step 15. \n",
    "\n",
    "<span style=\"color:blue\"> make.shared(list=BV5.trim.contigs.good.renamed.unique.pick.dgc.0.03.abund.list, count=BV5.trim.contigs.good.renamed.pick.0.03.abund.count_table)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
